File: apps/ios/Optio/Features/Live/Local/LocalAPI.swift
import Foundation
import SwiftUI

// MARK: - Response envelopes and route-local rows

/// Trigger attached to a local blueprint (`LocalTriggerSchema` in routes/local.ts).
/// Declared here because the route's row carries `targetType`/`targetId` and a
/// `type` that includes `ticket`, which the generated `WorkflowTrigger` lacks.
struct LocalTrigger: Decodable, Hashable, Identifiable {
    let id: String
    let targetType: String?
    let targetId: String?
    let type: String
    let config: [String: AnyCodable]?
    let paramMapping: [String: AnyCodable]?
    let enabled: Bool
    let lastFiredAt: Date?
    let nextFireAt: Date?
    let createdAt: Date?
    let updatedAt: Date?

    /// Mirrors `triggerSummary()` in blueprints-section.tsx.
    var summary: String {
        let cfg = config ?? [:]
        switch type {
        case "schedule":
            return cfg["cronExpression"]?.stringValue ?? ""
        case "webhook":
            return "/api/hooks/\(cfg["path"]?.stringValue ?? "")"
        case "ticket":
            let source = cfg["source"]?.stringValue ?? "any source"
            let labels = cfg["labels"]?.arrayValue?.compactMap(\.stringValue) ?? []
            return labels.isEmpty ? source : "\(source) · \(labels.joined(separator: ", "))"
        default:
            return ""
        }
    }

    var systemImage: String {
        switch type {
        case "schedule": return "clock"
        case "webhook": return "antenna.radiowaves.left.and.right"
        case "ticket": return "ticket"
        default: return "play"
        }
    }
}

// MARK: - Request bodies

struct CreateLocalTerminalBody: Encodable {
    var hostId: String
    var dir: String
    var title: String?
    var spec: LocalTerminalSpec
}

struct KillLocalTerminalBody: Encodable {
    var signal: String?
}

struct LocalTerminalInputBody: Encodable {
    var data: String
}

/// Shared by create (POST) and update (PATCH). Nil fields are omitted, matching
/// `blueprintBodySchema` (which has no nullable fields besides `agent`).
struct LocalBlueprintBody: Encodable {
    var name: String?
    var description: String?
    var hostId: String?
    var dir: String?
    var repoUrl: String?
    var commandTemplate: String?
    var agent: LocalAgentKind?
    /// Distinguishes "leave agent alone" (nil) from "explicitly shell" (true → null).
    var clearAgent = false
    var spawnMode: LocalBlueprintSpawnMode?
    var enabled: Bool?

    private enum CodingKeys: String, CodingKey {
        case name, description, hostId, dir, repoUrl, commandTemplate, agent, spawnMode, enabled
    }

    func encode(to encoder: any Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encodeIfPresent(name, forKey: .name)
        try c.encodeIfPresent(description, forKey: .description)
        try c.encodeIfPresent(hostId, forKey: .hostId)
        try c.encodeIfPresent(dir, forKey: .dir)
        try c.encodeIfPresent(repoUrl, forKey: .repoUrl)
        try c.encodeIfPresent(commandTemplate, forKey: .commandTemplate)
        if let agent {
            try c.encode(agent, forKey: .agent)
        } else if clearAgent {
            try c.encodeNil(forKey: .agent)
        }
        try c.encodeIfPresent(spawnMode, forKey: .spawnMode)
        try c.encodeIfPresent(enabled, forKey: .enabled)
    }
}

struct SpawnLocalBlueprintBody: Encodable {
    var params: [String: String] = [:]
}

struct CreateLocalTriggerBody: Encodable {
    var type: String
    var config: [String: AnyCodable]
    var enabled: Bool? = nil
}

// MARK: - Endpoints (mirror apps/web/src/lib/api-client.ts "Optio Local" block)

extension APIClient {
    private struct HostsEnvelope: Decodable { let hosts: [LocalHost] }
    private struct TerminalsEnvelope: Decodable { let terminals: [LocalTerminal] }
    private struct TerminalEnvelope: Decodable { let terminal: LocalTerminal }
    private struct BlueprintsEnvelope: Decodable { let blueprints: [LocalBlueprint] }
    private struct BlueprintEnvelope: Decodable { let blueprint: LocalBlueprint }
    private struct TriggersEnvelope: Decodable { let triggers: [LocalTrigger] }
    private struct TriggerEnvelope: Decodable { let trigger: LocalTrigger }

    func listLocalHosts() async throws -> [LocalHost] {
        try await get("/api/local/hosts", as: HostsEnvelope.self).hosts
    }

    func deleteLocalHost(_ id: String) async throws {
        try await delete("/api/local/hosts/\(id)")
    }

    func listLocalTerminals(hostId: String? = nil, state: String? = nil) async throws -> [LocalTerminal] {
        try await get("/api/local/terminals", query: ["hostId": hostId, "state": state], as: TerminalsEnvelope.self).terminals
    }

    func getLocalTerminal(_ id: String) async throws -> LocalTerminal {
        try await get("/api/local/terminals/\(id)", as: TerminalEnvelope.self).terminal
    }

    func createLocalTerminal(_ body: CreateLocalTerminalBody) async throws -> LocalTerminal {
        try await post("/api/local/terminals", body: body, as: TerminalEnvelope.self).terminal
    }

    func startLocalTerminal(_ id: String) async throws -> LocalTerminal {
        try await post("/api/local/terminals/\(id)/start", as: TerminalEnvelope.self).terminal
    }

    func killLocalTerminal(_ id: String, signal: String? = nil) async throws {
        try await post("/api/local/terminals/\(id)/kill", body: KillLocalTerminalBody(signal: signal))
    }

    /// REST fallback for stdin; the stream WS is the primary path.
    func sendLocalTerminalInput(_ id: String, data: String) async throws {
        try await post("/api/local/terminals/\(id)/input", body: LocalTerminalInputBody(data: data))
    }

    func deleteLocalTerminal(_ id: String) async throws {
        try await delete("/api/local/terminals/\(id)")
    }

    func listLocalBlueprints() async throws -> [LocalBlueprint] {
        try await get("/api/local/blueprints", as: BlueprintsEnvelope.self).blueprints
    }

    func getLocalBlueprint(_ id: String) async throws -> LocalBlueprint {
        try await get("/api/local/blueprints/\(id)", as: BlueprintEnvelope.self).blueprint
    }

    func createLocalBlueprint(_ body: LocalBlueprintBody) async throws -> LocalBlueprint {
        try await post("/api/local/blueprints", body: body, as: BlueprintEnvelope.self).blueprint
    }

    func updateLocalBlueprint(_ id: String, _ body: LocalBlueprintBody) async throws -> LocalBlueprint {
        try await patch("/api/local/blueprints/\(id)", body: body, as: BlueprintEnvelope.self).blueprint
    }

    func deleteLocalBlueprint(_ id: String) async throws {
        try await delete("/api/local/blueprints/\(id)")
    }

    func spawnLocalBlueprint(_ id: String, params: [String: String] = [:]) async throws -> LocalTerminal {
        try await post("/api/local/blueprints/\(id)/spawn", body: SpawnLocalBlueprintBody(params: params), as: TerminalEnvelope.self).terminal
    }

    func listLocalBlueprintTriggers(_ id: String) async throws -> [LocalTrigger] {
        try await get("/api/local/blueprints/\(id)/triggers", as: TriggersEnvelope.self).triggers
    }

    func createLocalBlueprintTrigger(_ id: String, _ body: CreateLocalTriggerBody) async throws -> LocalTrigger {
        try await post("/api/local/blueprints/\(id)/triggers", body: body, as: TriggerEnvelope.self).trigger
    }

    func deleteLocalBlueprintTrigger(_ id: String, triggerId: String) async throws {
        try await delete("/api/local/blueprints/\(id)/triggers/\(triggerId)")
    }
}

// MARK: - Presentation helpers (mirror terminal-card.tsx / work-links.tsx)

enum LocalPresentation {
    static let activeStates: Set<LocalTerminalState> = [.pending, .launching, .running]

    /// Last two path segments of an absolute dir — enough to recognize a checkout.
    static func dirTail(_ dir: String) -> String {
        let parts = dir.split(separator: "/").filter { !$0.isEmpty }
        let tail = parts.suffix(2).joined(separator: "/")
        return tail.isEmpty ? dir : tail
    }

    static func attentionLabel(_ reason: String?) -> String {
        switch reason {
        case "stop": return "waiting for you"
        case "notification": return "wants your attention"
        case "bell": return "rang the bell"
        case "quiet": return "gone quiet — probably waiting on you"
        case "exit": return "finished — review the result"
        default: return "needs you"
        }
    }

    static func stateLabel(_ t: LocalTerminal) -> String {
        if t.state == .pending, t.pendingReason == .hostOffline { return "Host offline" }
        if t.state == .pending, t.pendingReason == .hold { return "Held" }
        switch t.state {
        case .pending: return "Pending"
        case .launching: return "Launching"
        case .running: return "Running"
        case .exited: return "Exited"
        case .error: return "Error"
        case .unknown: return "Unknown"
        }
    }

    /// A live terminal that is waiting on the human: the daemon's `needs_you`, or a
    /// running terminal that has gone `idle` (an agent at its prompt in a plain shell,
    /// which the daemon cannot tell from a quiet command).
    static func waitsOnYou(_ t: LocalTerminal) -> Bool {
        guard !isDead(t) else { return false }
        if t.attentionState == .needsYou { return true }
        return t.state == .running && t.attentionState == .idle
    }

    /// Trailing label for a terminal that waits on you.
    static func waitingLabel(_ t: LocalTerminal) -> String {
        t.attentionState == .needsYou ? attentionLabel(t.attentionReason) : "waiting for input"
    }

    /// Terminal state → tone. Needs-you wins over everything while the process is alive.
    static func stateTone(_ t: LocalTerminal) -> Tone {
        if waitsOnYou(t) { return .accent }
        switch t.state {
        case .pending: return .idle
        case .launching, .running: return .working
        case .exited: return t.exitCode.map { $0 == 0 } ?? true ? .idle : .danger
        case .error: return .danger
        case .unknown: return .idle
        }
    }

    static func attentionTone(_ a: LocalAttentionState) -> Tone {
        switch a {
        case .needsYou: return .accent
        case .working: return .working
        case .idle, .unknown: return .idle
        }
    }

    /// Row dot: yellow while it waits on you, red on error, purple while working, none once finished.
    static func rowTone(_ t: LocalTerminal) -> Tone? {
        if waitsOnYou(t) { return .accent }
        if t.state == .error { return .danger }
        if t.state == .exited { return (t.exitCode ?? 0) == 0 ? nil : .danger }
        if t.state == .pending || t.state == .launching { return .idle }
        if t.attentionState == .working { return .working }
        return .idle
    }

    /// Ticket link merged in ahead of scanned links (`collectWorkLinks`).
    static func workLinks(_ t: LocalTerminal) -> [WorkLink] {
        let scanned = t.links
        guard let ticketUrl = t.ticketUrl, !scanned.contains(where: { $0.url == ticketUrl }) else { return scanned }
        let provider: WorkLinkProvider = t.ticketSource == "gitlab" ? .gitlab : .github
        let label = t.ticketExternalId.map { "#\($0)" } ?? "ticket"
        return [WorkLink(url: ticketUrl, kind: .issue, provider: provider, label: label)] + scanned
    }

    static func canStart(_ t: LocalTerminal) -> Bool { t.state == .pending && t.pendingReason != .hostOffline }
    static func canKill(_ t: LocalTerminal) -> Bool { t.state == .running || t.state == .launching }
    static func canDelete(_ t: LocalTerminal) -> Bool { t.state == .exited || t.state == .error || t.state == .pending }
    static func isDead(_ t: LocalTerminal) -> Bool { t.state == .exited || t.state == .error }

    static func spawnSourceIcon(_ s: LocalSpawnSource) -> String {
        switch s {
        case .manual: return "person"
        case .ticket: return "ticket"
        case .trigger: return "antenna.radiowaves.left.and.right"
        case .blueprint: return "square.stack.3d.up"
        case .api: return "cpu"
        case .unknown: return "questionmark"
        }
    }

    static func agentLabel(_ a: LocalAgentKind) -> String {
        switch a {
        case .claudeCode: return "Claude Code"
        case .codex: return "Codex"
        case .cursor: return "Cursor"
        case .gemini: return "Gemini"
        case .opencode: return "OpenCode"
        case .unknown: return a.rawValue
        }
    }

    /// Relative time from the row's activity timestamp (falls back to updatedAt).
    static func activityDescription(_ t: LocalTerminal) -> String {
        (t.lastActivityAt ?? t.updatedAt).relativeDescription
    }
}

/// Small chips for PR / ticket links (`WorkLinkBadges`). Taps open the URL.
struct WorkLinkBadges: View {
    let links: [WorkLink]
    var max: Int = 3

    /// `owner/repo#519` → `PR #519`; `group/proj!45` → `MR !45`; ticket refs (`ENG-12`) as-is.
    /// The repo is implied by the terminal's directory, so the pill only carries the number.
    static func shortLabel(_ link: WorkLink) -> String {
        let label = link.label
        guard let i = label.lastIndex(where: { $0 == "#" || $0 == "!" }) else { return label }
        let number = String(label[i...])
        switch link.kind {
        case .pr: return (number.hasPrefix("!") ? "MR " : "PR ") + number
        default: return number
        }
    }

    var body: some View {
        if !links.isEmpty {
            HStack(spacing: 4) {
                ForEach(links.prefix(max), id: \.url) { link in
                    if let url = URL(string: link.url) {
                        Link(destination: url) {
                            // Explicit HStack: `Label` collapses to icon-only inside a List row's Link.
                            HStack(spacing: 3) {
                                Image(systemName: link.kind == .pr ? "arrow.triangle.pull" : link.kind == .ref ? "number" : "circle.circle")
                                Text(Self.shortLabel(link))
                            }
                                .font(.caption2.monospaced())
                                .lineLimit(1)
                                .fixedSize()
                                .padding(.horizontal, 6)
                                .padding(.vertical, 2)
                                .background(.fill.tertiary, in: Radius.smallShape)
                                .foregroundStyle(.secondary)
                        }
                        .buttonStyle(.plain)
                    }
                }
                if links.count > max {
                    Text("+\(links.count - max)").font(.caption2.monospaced()).foregroundStyle(.tertiary)
                }
            }
        }
    }
}

File: apps/android/feature/local/src/main/kotlin/dev/optio/feature/local/api/LocalApi.kt
package dev.optio.feature.local.api

import dev.optio.core.model.LocalAgentKind
import dev.optio.core.model.LocalAgentSessionMode
import dev.optio.core.model.LocalBlueprint
import dev.optio.core.model.LocalBlueprintSpawnMode
import dev.optio.core.model.LocalHost
import dev.optio.core.model.LocalTerminal
import dev.optio.core.model.LocalTerminalSpec
import dev.optio.core.model.LocalTranscriptEntry
import dev.optio.core.network.ApiClient
import java.time.Instant
import kotlinx.serialization.Contextual
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonNull
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

// Endpoints of `apps/api/src/routes/local.ts`, as ApiClient extensions (iOS `LocalAPI.swift`, which
// mirrors the "Optio Local" block of the web's api-client.ts). Route-local envelopes live here.

// region Route-local rows and envelopes

/**
 * A trigger attached to a Local automation (`LocalTriggerSchema` in routes/local.ts). Declared here
 * because the route's row carries `targetType` / `targetId` and a `type` covering every trigger kind
 * (`schedule`, `webhook`, `ticket`, `github`, `slack`, `linear`, `manual`), which the generated
 * `WorkflowTrigger` doesn't model.
 */
@Serializable
data class LocalTrigger(
    val id: String,
    val targetType: String? = null,
    val targetId: String? = null,
    val type: String,
    val config: JsonObject? = null,
    val paramMapping: JsonObject? = null,
    val enabled: Boolean = true,
    @Contextual val lastFiredAt: Instant? = null,
    @Contextual val nextFireAt: Instant? = null,
    @Contextual val createdAt: Instant? = null,
    @Contextual val updatedAt: Instant? = null,
)

/**
 * `GET /api/local/terminals/:id/transcript`: one page of the stored conversation. [complete] is
 * true when fewer than `limit` entries came back, i.e. the caller has everything stored.
 */
@Serializable
data class LocalTranscriptPage(
    val entries: List<LocalTranscriptEntry> = emptyList(),
    val complete: Boolean = true,
)

@Serializable
internal data class HostsEnvelope(val hosts: List<LocalHost> = emptyList())

@Serializable
internal data class TerminalsEnvelope(val terminals: List<LocalTerminal> = emptyList())

@Serializable
internal data class TerminalEnvelope(val terminal: LocalTerminal)

@Serializable
internal data class BlueprintsEnvelope(val blueprints: List<LocalBlueprint> = emptyList())

@Serializable
internal data class BlueprintEnvelope(val blueprint: LocalBlueprint)

@Serializable
internal data class TriggersEnvelope(val triggers: List<LocalTrigger> = emptyList())

@Serializable
internal data class TriggerEnvelope(val trigger: LocalTrigger)

// endregion

// region Request bodies

@Serializable
data class CreateLocalTerminalBody(
    val hostId: String,
    val dir: String,
    val title: String? = null,
    val spec: LocalTerminalSpec = LocalTerminalSpec.Shell,
)

@Serializable
internal data class KillBody(val signal: String? = null)

@Serializable
internal data class InputBody(val data: String)

@Serializable
internal data class SnoozeBody(val minutes: Int)

@Serializable
internal data class RenameBody(val title: String)

@Serializable
internal data class SpawnBody(val params: Map<String, String> = emptyMap())

/**
 * An automation, for create (POST) and update (PATCH). Fields left null are omitted, matching
 * `blueprintBodySchema`. On a PATCH, [clearAgent] sends `agent: null` (a shell command), and
 * [clearLocation] sends `dir: null` / `repoUrl: null` for whichever of the two isn't set, so moving
 * an automation between "a directory", "a repo URL" and "the event's repo" really moves it (iOS's
 * `clearAgent`; the web editor sends the same explicit nulls).
 */
data class LocalBlueprintBody(
    val name: String? = null,
    val description: String? = null,
    val hostId: String? = null,
    val dir: String? = null,
    val repoUrl: String? = null,
    val commandTemplate: String? = null,
    val agent: LocalAgentKind? = null,
    val spawnMode: LocalBlueprintSpawnMode? = null,
    val sessionMode: LocalAgentSessionMode? = null,
    val enabled: Boolean? = null,
    val clearAgent: Boolean = false,
    val clearHost: Boolean = false,
    val clearDescription: Boolean = false,
    val clearLocation: Boolean = false,
) {
    /** The JSON the route takes: a map, so explicit nulls survive (`OptioJson` drops null properties). */
    fun toJson(): JsonObject {
        val out = LinkedHashMap<String, JsonElement>()
        name?.let { out["name"] = JsonPrimitive(it) }
        when {
            description != null -> out["description"] = JsonPrimitive(description)
            clearDescription -> out["description"] = JsonNull
        }
        when {
            hostId != null -> out["hostId"] = JsonPrimitive(hostId)
            clearHost -> out["hostId"] = JsonNull
        }
        when {
            dir != null -> out["dir"] = JsonPrimitive(dir)
            clearLocation -> out["dir"] = JsonNull
        }
        when {
            repoUrl != null -> out["repoUrl"] = JsonPrimitive(repoUrl)
            clearLocation -> out["repoUrl"] = JsonNull
        }
        commandTemplate?.let { out["commandTemplate"] = JsonPrimitive(it) }
        when {
            agent != null -> out["agent"] = JsonPrimitive(agent.raw)
            clearAgent -> out["agent"] = JsonNull
        }
        spawnMode?.let { out["spawnMode"] = JsonPrimitive(it.raw) }
        sessionMode?.let { out["sessionMode"] = JsonPrimitive(it.raw) }
        enabled?.let { out["enabled"] = JsonPrimitive(it) }
        return JsonObject(out)
    }
}

// endregion

// region Hosts

suspend fun ApiClient.listLocalHosts(): List<LocalHost> = get<HostsEnvelope>("/api/local/hosts").hosts

/** Unpairs a host (the server deletes its terminals too). */
suspend fun ApiClient.deleteLocalHost(id: String) = delete("/api/local/hosts/$id")

// endregion

// region Terminals

suspend fun ApiClient.listLocalTerminals(
    hostId: String? = null,
    state: String? = null,
): List<LocalTerminal> = get<TerminalsEnvelope>("/api/local/terminals", mapOf("hostId" to hostId, "state" to state)).terminals

suspend fun ApiClient.getLocalTerminal(id: String): LocalTerminal = get<TerminalEnvelope>("/api/local/terminals/$id").terminal

suspend fun ApiClient.createLocalTerminal(body: CreateLocalTerminalBody): LocalTerminal =
    post<TerminalEnvelope>("/api/local/terminals", body).terminal

/** Starts a `pending` (held) terminal. */
suspend fun ApiClient.startLocalTerminal(id: String): LocalTerminal = post<TerminalEnvelope>("/api/local/terminals/$id/start").terminal

/** SIGTERM by default; `SIGKILL` for "Force kill". */
suspend fun ApiClient.killLocalTerminal(
    id: String,
    signal: String? = null,
) = post("/api/local/terminals/$id/kill", KillBody(signal))

/** A new interactive terminal resuming the agent's own session (`claude --resume`); 409 without one. */
suspend fun ApiClient.resumeLocalTerminal(id: String): LocalTerminal =
    post<TerminalEnvelope>("/api/local/terminals/$id/resume", JsonObject(emptyMap())).terminal

/** "Later": out of the needs-you queue for [minutes] (server default 15, 1–1440). */
suspend fun ApiClient.snoozeLocalTerminal(
    id: String,
    minutes: Int,
): LocalTerminal = post<TerminalEnvelope>("/api/local/terminals/$id/snooze", SnoozeBody(minutes)).terminal

/** Clears a snooze so the terminal re-enters the needs-you queue. */
suspend fun ApiClient.unsnoozeLocalTerminal(id: String): LocalTerminal = delete<TerminalEnvelope>("/api/local/terminals/$id/snooze").terminal

suspend fun ApiClient.renameLocalTerminal(
    id: String,
    title: String,
): LocalTerminal = patch<TerminalEnvelope>("/api/local/terminals/$id", RenameBody(title)).terminal

/** REST fallback for stdin (the stream socket is the primary path); 409 unless running. */
suspend fun ApiClient.sendLocalTerminalInput(
    id: String,
    data: String,
) = post("/api/local/terminals/$id/input", InputBody(data))

/** Deletes a non-running terminal record. */
suspend fun ApiClient.deleteLocalTerminal(id: String) = delete("/api/local/terminals/$id")

/**
 * The conversation of an agent session (prompts, replies, tool calls), distilled by the daemon from
 * the agent CLI's own transcript. [after] fetches only entries past a seq.
 */
suspend fun ApiClient.getLocalTerminalTranscript(
    id: String,
    after: Long = 0,
    limit: Int = TRANSCRIPT_PAGE,
): LocalTranscriptPage =
    get<LocalTranscriptPage>(
        "/api/local/terminals/$id/transcript",
        mapOf("after" to after.takeIf { it > 0 }, "limit" to limit),
    )

/** The page size the web and iOS use. */
const val TRANSCRIPT_PAGE: Int = 2000

// endregion

// region Automations (blueprints)

suspend fun ApiClient.listLocalBlueprints(): List<LocalBlueprint> = get<BlueprintsEnvelope>("/api/local/blueprints").blueprints

suspend fun ApiClient.getLocalBlueprint(id: String): LocalBlueprint = get<BlueprintEnvelope>("/api/local/blueprints/$id").blueprint

suspend fun ApiClient.createLocalBlueprint(body: LocalBlueprintBody): LocalBlueprint =
    post<BlueprintEnvelope>("/api/local/blueprints", body.toJson()).blueprint

suspend fun ApiClient.updateLocalBlueprint(
    id: String,
    body: LocalBlueprintBody,
): LocalBlueprint = patch<BlueprintEnvelope>("/api/local/blueprints/$id", body.toJson()).blueprint

suspend fun ApiClient.deleteLocalBlueprint(id: String) = delete("/api/local/blueprints/$id")

/** A manual run ("Run now"): the terminal it spawned (pending when the automation holds). */
suspend fun ApiClient.spawnLocalBlueprint(
    id: String,
    params: Map<String, String> = emptyMap(),
): LocalTerminal = post<TerminalEnvelope>("/api/local/blueprints/$id/spawn", SpawnBody(params)).terminal

suspend fun ApiClient.listLocalBlueprintTriggers(id: String): List<LocalTrigger> =
    get<TriggersEnvelope>("/api/local/blueprints/$id/triggers").triggers

suspend fun ApiClient.createLocalBlueprintTrigger(
    id: String,
    type: String,
    config: JsonObject,
): LocalTrigger =
    post<TriggerEnvelope>(
        "/api/local/blueprints/$id/triggers",
        JsonObject(mapOf("type" to JsonPrimitive(type), "config" to config)),
    ).trigger

suspend fun ApiClient.setLocalBlueprintTriggerEnabled(
    id: String,
    triggerId: String,
    enabled: Boolean,
): LocalTrigger =
    patch<TriggerEnvelope>(
        "/api/local/blueprints/$id/triggers/$triggerId",
        JsonObject(mapOf("enabled" to JsonPrimitive(enabled))),
    ).trigger

suspend fun ApiClient.deleteLocalBlueprintTrigger(
    id: String,
    triggerId: String,
) = delete("/api/local/blueprints/$id/triggers/$triggerId")

// endregion
